### PR TITLE
Support access to CT log over DNS

### PR DESCRIPTION
**Pull request by daviddrysdale**
Somewhat experimental, might be easiest reviewed commit by commit.

### CODE REVIEW

File: client/logclient.go [hunk 1]
```diff
@@ -35,6 +35,13 @@ type LogClient struct {
 	jsonclient.JSONClient
 }
 
+// CheckLogClient is an interface that allows (just) checking of various log contents.
+type CheckLogClient interface {
```
Comment (Martin2112) [inline client/logclient.go]: Checking? or Verifying? Do we need "Log" as this is a log only package?

File: dnsclient/dnsclient.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+// DNSClient represents a DNS client for a given CT Log instance
+type DNSClient struct {
+	base     string
+	Verifier *ct.SignatureVerifier // nil for no verification (e.g. no public key available)
+	resolve  func(ctx context.Context, name string) ([]string, error)
+}
+
+// New constructs a new DNSClient instance.  The base parameter gives the
+// top-level domain name; opts can be used to provide a customer logger
```
Comment (Martin2112) [inline dnsclient/dnsclient.go]: typo: custom

File: dnsclient/dnsclient.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	}
+	return &DNSClient{
+		base:     base,
+		Verifier: verifier,
+		resolve:  func(ctx context.Context, name string) ([]string, error) { return net.LookupTXT(name) },
+	}, nil
+}
+
+const (
+	base64RE = "[A-Za-z0-9+/=]+"
```
Comment (Martin2112) [inline dnsclient/dnsclient.go]: I don't think this regex fully matches b64 as it allows '=' anywhere. Maybe that doesn't matter as false matches are unlikely in this case?

File: dnsclient/dnsclient.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+// GetSTH retrieves the current STH from the log.
+func (c *DNSClient) GetSTH(ctx context.Context) (*ct.SignedTreeHead, error) {
+	name := "sth." + c.base
+	results, err := c.resolve(ctx, name)
+	if err != nil {
+		return nil, fmt.Errorf("lookup for %q failed: %v", name, err)
+	}
+	result := strings.Join(results, "")
+	matches := sthTXT.FindStringSubmatch(result)
+	if matches == nil || len(matches) < 5 {
```
Comment (Martin2112) [inline dnsclient/dnsclient.go]: Magic number '5'.

File: dnsclient/dnsclient.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+	timestamp, err := strconv.ParseUint(matches[2], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse timestamp %q", matches[2])
+	}
+	rootHash, err := base64.StdEncoding.DecodeString(matches[3])
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse root hash %q", matches[3])
+	}
+	if len(rootHash) != sha256.Size {
+		return nil, fmt.Errorf("wrong size root hash %q", matches[3])
```
Comment (Martin2112) [inline dnsclient/dnsclient.go]: Should include both actual and expected values.

File: dnsclient/dnsclient.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+		TreeSize:  treeSize,
+		Timestamp: timestamp,
+	}
+	copy(sth.SHA256RootHash[:], rootHash)
+
+	var ds ct.DigitallySigned
+	if rest, err := tls.Unmarshal(signature, &ds); err != nil {
+		return nil, fmt.Errorf("failed to parse signature: %v", err)
+	} else if len(rest) > 0 {
+		return nil, fmt.Errorf("trailing data in parse signature: %v", err)
```
Comment (Martin2112) [inline dnsclient/dnsclient.go]: Isn't err nil here? Maybe a typo.

File: dnsclient/dnsclient.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		if err != nil {
+			if index == 0 {
+				return nil, fmt.Errorf("lookup for %q failed: %v", name, err)
+			}
+			// Assume that a failure to retrieve any more means the proof is complete.
+			break
+		}
+		result := []byte(strings.Join(results, ""))
+		// We expect the result to be a concatenation of hashes, and so should be a multiple of the hash size.
+		if len(result)%sha256.Size != 0 {
```
Comment (Martin2112) [inline dnsclient/dnsclient.go]: Does it do the right thing if len(result) == 0?

File: client/ctclient/ctclient.go [hunk 8]
```diff
@@ -342,29 +345,47 @@ func main() {
 		}
 		opts.PublicKey = string(pubkey)
 	}
-	logClient, err := client.New(*logURI, httpClient, opts)
+
+	var err error
+	var logClient *client.LogClient
+	var checkClient client.CheckLogClient
+	if *dnsBase != "" {
```
Comment (Martin2112) [inline client/ctclient/ctclient.go]: Could we make checkClient a logClient that rejects the stuff it can't do? Then this initialization and the checks below would be simpler.

File: client/logclient.go (hunk 1 above)
Comment (Martin2112) [inline client/logclient.go]: As noted elsewhere if we didn't have this interface and the DNS code returned not implemented errors then it could make some things simpler.

Comment (daviddrysdale) [reply client/logclient.go]: It's analogous to the existing `LogClient` and `MultiLogClient` structs and the `AddLogClient` interface, so `CheckLogClient` was the best I could come up with.

File: dnsclient/dnsclient.go (hunk 2 above)
Comment (daviddrysdale) [reply dnsclient/dnsclient.go]: Fixed here and at its [origin](https://github.com/google/certificate-transparency-go/commit/94d42246b1a9bfff17569e).

File: dnsclient/dnsclient.go (hunk 3 above)
Comment (daviddrysdale) [reply dnsclient/dnsclient.go]: It shouldn't matter it's just for finding substring boundaries (the decode step later will check it), but added a comment.

File: dnsclient/dnsclient.go (hunk 4 above)
Comment (daviddrysdale) [reply dnsclient/dnsclient.go]: Done.

File: dnsclient/dnsclient.go (hunk 5 above)
Comment (daviddrysdale) [reply dnsclient/dnsclient.go]: Done.

File: dnsclient/dnsclient.go (hunk 6 above)
Comment (daviddrysdale) [reply dnsclient/dnsclient.go]: Yeah, typo; fixed.

File: dnsclient/dnsclient.go (hunk 7 above)
Comment (daviddrysdale) [reply dnsclient/dnsclient.go]: Good spot -- the old first==last edge case.  Added a check and a test.

File: client/ctclient/ctclient.go (hunk 8 above)
Comment (daviddrysdale) [reply client/ctclient/ctclient.go]: I think it's better to catch things at compile-time than at run-time whenever possible, so I'm not keen on the `unimplemented` approach.  (But then, I've been known to write Ocaml for fun...)

Comment (Martin2112) [reply client/ctclient/ctclient.go]: This is caught at runtime though? If you set options for DNS and something it doesn't support.

Comment (daviddrysdale) [reply client/ctclient/ctclient.go]: Maybe I misunderstood - I thought you were suggesting making the `DNSClient` implement the full `LogClient` interface, then removing the operation-specific checks and letting the `errUnimplemented` flow out to the user.

Comment (Martin2112) [reply client/ctclient/ctclient.go]: Yes, that's what I meant. It doesn't seem that different and removes extra logic.

File: dnsclient/dnsclient.go [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+		base:     base,
+		Verifier: verifier,
+		resolve:  func(ctx context.Context, name string) ([]string, error) { return net.LookupTXT(name) },
+	}, nil
+}
+
+const (
+	// Match base64 data (although note that this will also match invalid data with
+	// padding in the middle rather than the end).
+	base64RE = "[A-Za-z0-9+/=]+"
```
Comment (pphaneuf) [inline dnsclient/dnsclient.go]: Wouldn't be too difficult to change the regex to accept the '=' character only at the end? Just a suggestion, feel free to ignore. :stuck_out_tongue:

File: dnsclient/dnsclient.go [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+	// DNS results for get-sth have 5 matches: 1 for the overall match, plus 4
+	// dot-delimited fields
+	sthMatchCount = 1 + 4
+)
+
+var (
+	sthTXT = regexp.MustCompile(`^(\d+)\.(\d+)\.(` + base64RE + `)\.(` + base64RE + `)`)
```
Comment (pphaneuf) [inline dnsclient/dnsclient.go]: Okay, maybe that's a good reason why the previous comment might be more difficult than it looks... :laughing:
